### ISSUE TITLE
React useTable hook with typescript

### ISSUE THREAD

**Issue by godomainz** [TanStack/react-table#3064]
So I have a JavaScript class with useTable. In Java script its been used like below

```
import {useTable, useFilters,useAsyncDebounce,useSortBy,usePagination,} from "react-table";

const {getTableProps,getTableBodyProps, headerGroups, page,prepareRow,state,visibleColumns, nextPage,pageOptions,pageCount,previousPage,canPreviousPage,canNextPage,setPageSize,gotoPage
      } = useTable({columns,data,defaultColumn,filterTypes,initialState: { pageSize: 10, pageIndex: 0 }
       },useFilters, useSortBy,usePagination);
```
I'm converting my javascript project to typescript.
In typescript that `useTable ` line gives errors.
Example : `Property 'page' does not exist on type 'TableInstance<object>'.ts(2339)`

This is a ScreenShot of my errors: [https://i.stack.imgur.com/taGkw.png](url)

Basically all of these attributes gives me error -> page, nextPage, pageOptions, pageCount, previousPage, canPreviousPage, canNextPage, setPageSize, gotoPage, initialState

That `initialState `attribute gives my below error

```
Type '{ pageSize: number; pageIndex: number; }' is not assignable to type 'Partial<TableState<object>>'.
  Object literal may only specify known properties, and 'pageSize' does not exist in type 'Partial<TableState<object>>'.ts(2322)
index.d.ts(172, 5): The expected type comes from property 'initialState' which is declared here on type 'TableOptions<object>'
```
How to write that `useTable `line in Typescript ?

**Comment by avinoamsn** [comment]
No one here is going to have an adequate solution to your issue, but a quick google/issue search would probably do wonders. If you check out https://github.com/tannerlinsley/react-table/issues/1591, you'll see that there's been an ongoing discussion regarding the TS deficiencies that plague this lib. If you read through the thread, you can see that the author of the type defs is no longer actively updating the `@types` lib https://github.com/tannerlinsley/react-table/pull/1534#issuecomment-560533319, and that the primary maintainer of this repo is currently rewriting the lib to natively use TypeScript for the v8 release https://github.com/tannerlinsley/react-table/issues/1591#issuecomment-729945218.

TLDR; the latest release of `react-table` doesn't have up-to-date type defs, and there's no ideal solution to this at the moment.

That being said, you can implement a number of workarounds to address your problem in the short term. The easiest option is just to typecast `useTable`'s type to `any` (`useTable(...) as any`), which'll obliterate any TS intellisense but it's the fastest way to alleviate the compilation problems you have (and if the table has already been implemented in JS, then you probably don't need any of the lift TS provides). A better solution would be to modify your compiler config to ignore these warnings & transpile to JS even if the types are broken (this solution won't break TS intellisense & you'll still have linter hints in your editor). A third option is to use the babel compiler `babel-loader` w/ the TS preset `@babel/preset-typescript`, instead of the TS compiler `ts-loader` (this option won't affect your development env, but it will mean that TS issues aren't recognized at compile time b/c babel's loader doesn't have type def functionality). This final option is how my current project is configured.

**Comment by Pavinati** [comment]
Instead of casting to `any` i extended the table instance type

```ts
import { useTable, usePagination } from "react-table";
import type { TableInstance } from "react-table";

// Required workaround for missing TypesScript definitions.
// Will be fixed in react-table v8
// see also https://github.com/tannerlinsley/react-table/issues/3064
type TableTypeWorkaround<T extends Object> = TableInstance<T> & {
  gotoPage: (index: number) => void;
  state: {
    pageIndex: number;
    pageSize: number;
  }
};

[...]

  const tableInstance = useTable({ columns, data }, usePagination) as TableTypeWorkaround<T>;
```

It requires some TypeScript foo, but you won't lose the useful typechecking.

**Comment by jstafman** [comment]
Similar to @Pavinati, I extended the table instance as well, but I used other types already defined by react-table.

```javascript
import {
  TableInstance,
  UsePaginationInstanceProps,
  UsePaginationState,
} from 'react-table';

export type PaginationTableInstance<T extends object> = TableInstance<T> &
  UsePaginationInstanceProps<T> & {
    state: UsePaginationState<T>;
  };

```

**Comment by danielweck** [comment]
I must admit being very surprised (sadly) to discover the typing "holes" in `[@types/]react-table`, but I also appreciate how good this library is in terms of its features :) So thank you for that 👍 

I will put in the extra work to work around the TypeScript compiler errors, and I look forward to trying the next major version of `react-table` written in TS from scratch! (currently in early preview with bugs and missing features, so a no-go for my project(s))

**Comment by tannerlinsley** [comment]
v8 is written entirely in Typescript, so I'm going to call this issue "fixed" from here on out.
